internal/v2/tls_config_store/tls_config_store.go

```diff
@@ -21,54 +33,242 @@ var (
+// a helper function.
+
 // GetTlsConfigurationForClient returns a tls.Config instance for use by a client application.
 func GetTlsConfigurationForClient(serverHostname string) *tls.Config {
-	// TODO(rmehta19): Call S2Av2 for certificate.
-	// TODO(rmehta19): Call remote signer library for private key.
-	cert, err := tls.X509KeyPair(clientCert, clientKey)
+	// Create stream to S2Av2.
+	opts := []grpc.DialOption {
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithReturnConnectionError(),
+		grpc.WithBlock(),
+	}
+	conn, err := grpc.Dial(*fakes2av2Address, opts...)
+	if err != nil {
+		log.Fatalf("Client: failed to connect: %v", err)
+	}
+	defer conn.Close()
+	c := s2av2pb.NewS2AServiceClient(conn)
+	log.Printf("Client: connected to: %s", *fakes2av2Address)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+	defer cancel()
+
+	// Setup bidrectional streaming session.
+	callOpts := []grpc.CallOption{}
+	cstream, err := c.SetUpSession(ctx, callOpts...)
+	if err != nil  {
+		log.Fatalf("Client: failed to setup bidirectional streaming RPC session: %v", err)
+	}
+	log.Printf("Client: set up bidirectional streaming RPC session.")
+
+	// Send request to S2Av2 for config.
+	err = cstream.Send(&s2av2pb.SessionReq {
+		AuthenticationMechanisms: []*s2av2pb.AuthenticationMechanism {
+			{
+				MechanismOneof: &s2av2pb.AuthenticationMechanism_Token{"token"},
+			},
+		},
+		ReqOneof: &s2av2pb.SessionReq_GetTlsConfigurationReq {
+			&s2av2pb.GetTlsConfigurationReq {
+				ConnectionSide: commonpb.ConnectionSide_CONNECTION_SIDE_CLIENT,
+			},
+		},
+	})
 	if err != nil {
-		log.Fatalf("Failed to generate X509KeyPair: %v", err)
+		log.Fatalf("Client: failed to send SessionReq: %v", err)
+	}
+	log.Printf("Client: sent SessionReq")
+
+	// Get the response containing config from S2Av2.
+	resp, err := cstream.Recv()
+	if err != nil {
+		log.Fatalf("Client: failed to recieve SessionResp: %v", err)
+	}
+	log.Printf("Client: recieved SessionResp")
+
+	// TODO(rmehta19): Handle resp.GetStatus().
+
+	// Extract TLS configiguration from SessionResp.
+	tlsConfig := resp.GetGetTlsConfigurationResp().GetClientTlsConfiguration()
+
+	// Fake S2Av2 only puts one cert in CertitificateChain, for loop iterates
+	// once.
+	var certList []tls.Certificate
+	for _, v := range tlsConfig.CertificateChain {
+		// TODO(rmehta19): Call remote signer library for private key.
+		cert, err := tls.X509KeyPair([]byte(v), clientKey)
+		if err != nil {
+			log.Fatalf("Failed to generate X509KeyPair: %v", err)
+		}
+		certList = append(certList, cert)
 	}
 
 	rootCertPool := x509.NewCertPool()
 	rootCertPool.AppendCertsFromPEM(serverCert)
 
-	// TODO(rmehta19): Call S2Av2 for config values.
+	// Map S2Av2 TLSVersion to consts defined in tls package.
+	var minVersion uint16
+	var maxVersion uint16
+	switch x := tlsConfig.MinTlsVersion; x {
+	case commonpb.TLSVersion_TLS_VERSION_UNSPECIFIED:
```

Comment:
Should we omit the "unspecified" case and just let it fall into the default bucket?

Reply:
Good idea, thanks!